File: Algorithms/Quick_Sort_Complexity_Estimation.f90
!===========================================================
! QuickSort Algorithm with Empirical Complexity Measurement
! Language: Fortran 90+
! gfortran Quick_Sort_Complexity_Estimation.f90 -o quicksort
! ./quicksort
!===========================================================

program quicksort_complexity
    implicit none
    integer :: n, i
    real :: start_time, end_time, elapsed_time
    real :: theo_complexity, ratio
    integer, allocatable :: arr(:)
    integer(kind=8) :: comparisons, swaps

    comparisons = 0
    swaps = 0

    print *, "Enter number of elements:"
    read(*,*) n
    allocate(arr(n))

    print *, "Enter ", n, " integers:"
    read(*,*) arr

    call cpu_time(start_time)
    call quicksort(arr, 1, n, comparisons, swaps)
    call cpu_time(end_time)

    elapsed_time = end_time - start_time
    theo_complexity = n * log(real(n)) / log(2.0)
    ratio = real(comparisons) / theo_complexity

    print *, ""
    print *, "Sorted array:"
    do i = 1, n
        write(*,'(I8)', advance='no') arr(i)
    end do
    print *, ""

    print *, ""
    print *, "=== QuickSort Complexity Report ==="
    print *, "Number of elements (n): ", n
    print *, "Comparisons: ", comparisons
    print *, "Swaps: ", swaps
    print *, "Execution time (sec): ", elapsed_time
    print *, "Estimated O(n log n): ", theo_complexity
    print *, "Empirical ratio (comparisons / n log n): ", ratio

    deallocate(arr)
end program quicksort_complexity

!===========================================================
contains
!===========================================================

recursive subroutine quicksort(arr, low, high, comparisons, swaps)
    implicit none
    integer, intent(inout) :: arr(:)
    integer, intent(in) :: low, high
    integer(kind=8), intent(inout) :: comparisons, swaps
    integer :: pivot_index

    if (low < high) then
        call partition(arr, low, high, pivot_index, comparisons, swaps)
        call quicksort(arr, low, pivot_index - 1, comparisons, swaps)
        call quicksort(arr, pivot_index + 1, high, comparisons, swaps)
    end if
end subroutine quicksort

!-----------------------------------------------------------
subroutine partition(arr, low, high, pivot_index, comparisons, swaps)
    implicit none
    integer, intent(inout) :: arr(:)
    integer, intent(in) :: low, high
    integer, intent(out) :: pivot_index
    integer(kind=8), intent(inout) :: comparisons, swaps

    integer :: pivot, i, j, temp

    pivot = arr(high)
    i = low - 1

    do j = low, high - 1
        comparisons = comparisons + 1
        if (arr(j) <= pivot) then
            i = i + 1
            call swap(arr(i), arr(j), swaps)
        end if
    end do
    call swap(arr(i + 1), arr(high), swaps)
    pivot_index = i + 1
end subroutine partition

!-----------------------------------------------------------
subroutine swap(a, b, swaps)
    implicit none
    integer, intent(inout) :: a, b
    integer(kind=8), intent(inout) :: swaps
    integer :: temp

    temp = a
    a = b
    b = temp
    swaps = swaps + 1
end subroutine swap
!===========================================================

File: Algorithms/Quick_Sort_Complexity_Estimation.c
// QuickSort with Empirical Complexity Estimation
// Description: Reads numbers, sorts them using QuickSort,
// counts comparisons and swaps, and prints empirical complexity.

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

long long comparisons = 0; // global counter for comparisons
long long swaps = 0;       // global counter for swaps

// Swap two elements
void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
    swaps++;
}

// Partition function (Lomuto partition scheme)
int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    for (int j = low; j < high; j++) {
        comparisons++;
        if (arr[j] <= pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return i + 1;
}

// Recursive QuickSort
void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

// Function to print array
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    int n;
    printf("Enter number of elements: ");
    scanf("%d", &n);

    int *arr = (int *)malloc(n * sizeof(int));
    if (!arr) {
        printf("Memory allocation failed.\n");
        return 1;
    }

    printf("Enter %d integers:\n", n);
    for (int i = 0; i < n; i++)
        scanf("%d", &arr[i]);

    clock_t start = clock();
    quickSort(arr, 0, n - 1);
    clock_t end = clock();

    double time_taken = ((double)(end - start)) / CLOCKS_PER_SEC;

    printf("\nSorted array:\n");
    printArray(arr, n);

    printf("\n=== QuickSort Complexity Report ===\n");
    printf("Number of elements (n): %d\n", n);
    printf("Comparisons: %lld\n", comparisons);
    printf("Swaps: %lld\n", swaps);
    printf("Execution time: %.6f seconds\n", time_taken);
    printf("Estimated O(n log n): %.2f\n", n * (log(n) / log(2)));
    printf("Empirical ratio (comparisons / n log n): %.2f\n",
           comparisons / (n * (log(n) / log(2))));

    free(arr);
    return 0;
}
